core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala

```diff
@@ -0,0 +1,269 @@
+
+import kafka.common.{KafkaException, MessageFormatter}
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.protocol.types.Type._
+import org.apache.kafka.common.protocol.types.{ArrayOf, Field, Schema, Struct}
+import java.io.PrintStream
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+
+import org.apache.kafka.common.record.CompressionType
+
+import scala.collection.mutable
+
+/*
+ * Messages stored for the transaction topic represent the pid and transactional status of the corresponding
+ * transactional id, which have versions for both the key and value fields. Key and value
+ * versions are used to evolve the message formats:
+ *
+ * key version 0:               [transactionalId]
+ *    -> value version 0:       [pid, epoch, expire_timestamp, status, [topic [partition] ]
+ */
+object TransactionLog {
+
+  // log-level config default values and enforced values
+  val DefaultNumPartitions: Int = 50
+  val DefaultSegmentBytes: Int = 100 * 1024 * 1024
+  val DefaultReplicationFactor: Short = 3.toShort
+  val DefaultMinInSyncReplicas: Int = 2
+  val DefaultLoadBufferSize: Int = 5 * 1024 * 1024
+
+  // enforce always using
+  //  1. cleanup policy = compact
+  //  2. compression = none
+  //  3. unclean leader election = disabled
+  //  4. required acks = -1 when writing
+  val EnforcedCompressionType: CompressionType = CompressionType.NONE
+  val EnforcedRequiredAcks: Short = (-1).toShort
+
+  // log message formats
+  private val TXN_ID_KEY = "transactional_id"
+
+  private val PID_KEY = "pid"
+  private val EPOCH_KEY = "epoch"
+  private val TXN_TIMEOUT_KEY = "transaction_timeout"
+  private val TXN_STATUS_KEY = "transaction_status"
+  private val TXN_PARTITIONS_KEY = "transaction_partitions"
+  private val TXN_TIMESTAMP_FIELD = "transaction_timestamp"
+  private val TOPIC_KEY = "topic"
+  private val PARTITION_IDS_KEY = "partition_ids"
+
+  private val KEY_SCHEMA_V0 = new Schema(new Field(TXN_ID_KEY, STRING))
+  private val KEY_SCHEMA_TXN_ID_FIELD = KEY_SCHEMA_V0.get(TXN_ID_KEY)
+
+  private val VALUE_PARTITIONS_SCHEMA = new Schema(new Field(TOPIC_KEY, STRING),
+                                                   new Field(PARTITION_IDS_KEY, new ArrayOf(INT32)))
+  private val PARTITIONS_SCHEMA_TOPIC_FIELD = VALUE_PARTITIONS_SCHEMA.get(TOPIC_KEY)
+  private val PARTITIONS_SCHEMA_PARTITION_IDS_FIELD = VALUE_PARTITIONS_SCHEMA.get(PARTITION_IDS_KEY)
+
+  private val VALUE_SCHEMA_V0 = new Schema(new Field(PID_KEY, INT64),
+                                           new Field(EPOCH_KEY, INT16),
+                                           new Field(TXN_TIMEOUT_KEY, INT32),
+                                           new Field(TXN_STATUS_KEY, INT8),
+                                           new Field(TXN_PARTITIONS_KEY, ArrayOf.nullable(VALUE_PARTITIONS_SCHEMA)),
+                                           new Field(TXN_TIMESTAMP_FIELD, INT64))
```

Comment:
The comment in line 38 doesn't include TXN_TIMESTAMP_FIELD. Also, could we add a doc for each field?